source/MRMesh/MRMarchingCubes.cpp

```diff
@@ -499,28 +513,21 @@ Expected<TriMesh> volumeToMesh( const V& volume, const MarchingCubesParams& para
             return;
         const auto layerEnd = std::min( ( blockIndex + 1 ) * layerPerBlockCount, layerCount - 1 ); // skip last layer since no data from next layer
 
-        const VoxelsVolumeAccessor<V> acc( volume );
-        std::optional<VoxelsVolumeCachingAccessor<V>> cache;
-        if ( cachingMode == MarchingCubesParams::CachingMode::Normal )
-        {
-            using Parameters = typename VoxelsVolumeCachingAccessor<V>::Parameters;
-            cache.emplace( acc, indexer, Parameters {
-                .preloadedLayerCount = 2,
-            } );
-            cache->preloadLayer( layerBegin );
-        }
-
         // cell data
         std::array<const SeparationPointSet*, 7> neis;
         unsigned char voxelConfiguration;
         VoxelLocation loc = indexer.toLoc( Vector3i( 0, 0, layerBegin ) );
         for ( ; loc.pos.z < layerEnd; ++loc.pos.z )
         {
-            if ( cache && loc.pos.z != cache->currentLayer() )
+            const BitSet* layerInvalids[2] = { &invalids[loc.pos.z], &invalids[loc.pos.z+1] };
+            const BitSet* layerLowerIso[2] = { &lowerIso[loc.pos.z], &lowerIso[loc.pos.z+1] };
+            const VoxelId firstLayerVoxelId[2] = { indexer.toVoxelId( { 0, 0, loc.pos.z } ), indexer.toVoxelId( { 0, 0, loc.pos.z + 1 } ) };
+            auto getBit = [&]( const BitSet *bs[2], const VoxelLocation & vl )
             {
-                cache->preloadNextLayer();
-                assert( loc.pos.z == cache->currentLayer() );
-            }
+                const auto dl = vl.pos.z - loc.pos.z;
+                assert( dl >= 0 && dl <= 1 );
+                return (*bs)[dl][vl.id - firstLayerVoxelId[dl]];
+            };
```

Comment:
This section looks hard to read. Maybe better to make a separate helper class with more detailed description.